### PR TITLE
Health connect fixes

### PR DESCRIPTION
**Pull request by Aozora7**
Based on feedback in #586

* Removed the intent to open Play Store to install Health Connect from the foss flavor. gplay flavor will open Play Store, while foss flavor will disable the button and show text prompting the user to install the provider manually. This only affects devices with Android 9-13, since from 14 onwards Health Connect is a part of the system. Tell me if this doesn't alleviate your concern and I will move the entire health connect feature to gplay only as a part of this PR. 
* Moved Health Connect UUID assignment from sleep object initialization to Health Connect write. This change includes not writing deletions when there is no health connect ID in the record. This required removing unique from the index, I added the change to the existing migration since this has not been released yet. 
* Adjusted reconcileForeground to return early if there are no writes or deletions pending. There can still be edge cases where the user might want to run the full sync without pending writes, added a separate settings button for that specifically. 
* Removed the version-specific strings file
* Split test cases to avoid execution after assertions. 
* Checked empty metadata round-trip on device. Work correctly. Added tests for null and empty string. 
* Updated copyright notices to use my real name, it's public anyway. 
* Added comments to health connect database fields to explain their purpose/necessity.

> health_connect_version is exported to CSV, do you want to export health_connect_synced_version to CSV as well?

health_connect_synced_version is not exported to CSV since it would provide no useful information when imported. If the app exported a CSV with a sleep record that still had a pending update for health connect on the current device, it would still export the newest version, and so the importing app would also get the newest version and write that version to health connect if enabled. 

I attempted to address everything actionable. Please remind me if I missed anything.

### CODE REVIEW

File: app/src/main/java/hu/vmiklos/plees_tracker/Sleep.kt [hunk 1]
```diff
@@ -38,7 +37,7 @@ class Sleep {
 
     // Identifies the record across devices, where sid values can collide.
     @ColumnInfo(name = "health_connect_id")
-    var healthConnectId: String = UUID.randomUUID().toString()
+    var healthConnectId: String = ""
```
Comment (vmiklos) [inline app/src/main/java/hu/vmiklos/plees_tracker/Sleep.kt:40]: Thanks. I asked for this.

File: app/src/main/res/values-v34/strings.xml [hunk 2]
```diff
@@ -1,6 +0,0 @@
-<resources>
```
Comment (vmiklos) [inline app/src/main/res/values-v34/strings.xml:1]: Again I asked for this, thanks.

File: app/src/gplay/java/hu/vmiklos/plees_tracker/HealthConnectProviderInstaller.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+/** Opens Google Play to install or update the Health Connect provider. */
+object HealthConnectProviderInstaller {
+    private const val PROVIDER_PACKAGE_NAME = "com.google.android.apps.healthdata"
+    private const val PLAY_STORE_PACKAGE_NAME = "com.android.vending"
+    private const val PLAY_STORE_URL =
+        "https://play.google.com/store/apps/details?id=$PROVIDER_PACKAGE_NAME"
+
+    // Not const so shared code does not trigger constant-condition warnings.
+    val isSupported = true
```
Comment (vmiklos) [inline app/src/gplay/java/hu/vmiklos/plees_tracker/HealthConnectProviderInstaller.kt:20]: I asked for this, thanks.

File: app/src/main/java/hu/vmiklos/plees_tracker/HealthConnectBackend.kt [hunk 4]
```diff
@@ -346,14 +350,18 @@ object HealthConnectBackend {
...
+        val pending = validSleeps(DataModel.database.sleepDao().getPendingHealthConnectWrites())
+        write(client, pending)
     }
 
     @RequiresApi(Build.VERSION_CODES.P)
     private suspend fun write(client: HealthConnectClient, sleeps: List<Sleep>) {
         val healthDao = DataModel.database.healthConnectDao()
-        for (chunk in sleeps.chunked(PAGE_SIZE)) {
+        for (sleep in sleeps.filter { it.healthConnectId.isEmpty() }) {
+            sleep.healthConnectId = UUID.randomUUID().toString()
```
Comment (vmiklos) [inline app/src/main/java/hu/vmiklos/plees_tracker/HealthConnectBackend.kt:361]: Hm, is it correct to assign this UUID this late, during sync? This runs on a thread, and the UI code at DataModel.deleteSleep() seems to have an 'if (sleep.healthConnectId.isNotEmpty())' condition to decide if insertDeletions() is called. So if you're unlucky, the UI's delete action is called between the sync thread's getPendingHealthConnectWrites() and insertRecords() and the sleep is deleted, but insertDeletions() won't be called.

Maybe you could solve this by making sure the UUID generation is called when the UI deletes the sleep, the file import / calendar import / health connect import creates the sleep, so there is no cross-thread race here.

Extra bonus points: this codebase uses coroutines for async work, mostly. And then you introduced operationMutex -- it would be ideal to convert the health connect sync code to a coroutine as well, at least I find explicit threading hard to reason about, data races are very easy to make.

File: app/src/main/java/hu/vmiklos/plees_tracker/HealthConnectDao.kt [hunk 5]
```diff
@@ -29,6 +32,9 @@ interface HealthConnectDao {
         }
     }
 
+    @Query("UPDATE sleep SET health_connect_id = :id WHERE sid = :sid AND health_connect_id = ''")
+    suspend fun assignId(sid: Int, id: String)
```
Comment (vmiklos) [inline app/src/main/java/hu/vmiklos/plees_tracker/HealthConnectDao.kt:36]: Is it a problem that if the "where" conditions are not met here, then once assignId() finishes, the in-memory object and what sql has differs?

File: app/src/main/java/hu/vmiklos/plees_tracker/Preferences.kt [hunk 6]
```diff
@@ -127,6 +127,18 @@ class Preferences : PreferenceFragmentCompat() {
...
             category.addPreference(this)
         }
+        Preference(requireContext()).apply {
+            key = PreferencesActivity.HEALTH_CONNECT_FORCE_SYNC_KEY
+            order = 2
+            setTitle(R.string.settings_health_connect_force_sync)
+            setSummary(R.string.settings_health_connect_force_sync_summary)
+            isVisible = false
+            setOnPreferenceClickListener {
+                (activity as? PreferencesActivity)?.forceHealthConnectSync()
```
Comment (vmiklos) [inline app/src/main/java/hu/vmiklos/plees_tracker/Preferences.kt:137]: Do you want to only allow one forced sync at a time?

**Review by vmiklos — APPROVED** [review]
Looks OK, thanks for these. See inline comments for some further ideas if you are interested; but overall the direction looks correct and most of it is clean.

Last nit: please increase the schema version if you touch it again, the schema before and after this PR is not the same and the version was not incremented. (No need to fix it just now, I note this for the future).

File: app/src/main/java/hu/vmiklos/plees_tracker/HealthConnectBackend.kt (hunk 4 above)
Comment (Aozora7) [reply app/src/main/java/hu/vmiklos/plees_tracker/HealthConnectBackend.kt:361]: I think the narrow fix would be to retrieve the record by sid in deleteSleep to ensure insertDeletions gets run if it's needed. I think assigning UUID during write is reasonable otherwise since this is the only place when they start needing it. 

When it comes to refactoring HealthConnectBackend to use coroutines, what would that look like? Storing a queue of coroutines and executing them sequentially in the order they were called?

Comment (vmiklos) [reply app/src/main/java/hu/vmiklos/plees_tracker/HealthConnectBackend.kt:361]: > I think assigning UUID during write is reasonable otherwise since this is the only place when they start needing it.

OK, fair enough.

> When it comes to refactoring HealthConnectBackend to use coroutines, what would that look like? Storing a queue of coroutines and executing them sequentially in the order they were called?

Ah sorry, somehow I missed that your mutex is a kotlinx.coroutines.sync.Mutex, so you already don't use threads here, good. My thinking was that you could avoid this mutex as well if you:

1) make WorkManager the only entry point
2) the WorkManager already serializes execution by unique name, it seems to me
3) scheduleSync uses WORK_NAME, so if you enqueue the work instead of calling reconcileForeground directly, then WorkManager would automatically make sure that only one job is running in parallel and you don't need the mutex

Would this work or did I miss something? I attach an untested prototype -- it would be interesting if you could test it still works. If so, this would be a cleaner design, I would say. What do you think?

[patch.txt](https://github.com/user-attachments/files/30673535/patch.txt)

Comment (Aozora7) [reply app/src/main/java/hu/vmiklos/plees_tracker/HealthConnectBackend.kt:361]: I do think your implementation is easier to understand. 

I tested it manually, with existing on-device tests, and a couple of new on-device tests for cancellation and multiple inserts/updates in a short time frame. Everything passes. 

Should I clean this up, add unit tests, data race fix, and make a new PR with that?

Comment (vmiklos) [reply app/src/main/java/hu/vmiklos/plees_tracker/HealthConnectBackend.kt:361]: Please do. :-) My patch was just to demonstrate the idea. Thanks.
